amy/workshops/models.py

```diff
@@ -2432,35 +2603,96 @@ class TrainingProgress(CreatedUpdatedMixin, models.Model):
         on_delete=models.SET_NULL,
     )
     url = models.URLField(null=True, blank=True, verbose_name="URL")
+    trainee_notes = models.CharField(
+        blank=True, max_length=STR_LONGEST, verbose_name="Notes from trainee"
+    )
     notes = models.TextField(blank=True)
 
     def get_absolute_url(self):
         return reverse("trainingprogress_edit", args=[str(self.id)])
 
     def clean(self):
```

Comment:
This has become crazy complex. I'm wondering if we should perhaps introduce some sort of strategy pattern for this validation.

Reply:
Did you cover all the cases in unit tests?

Reply:
I agree that there's room for improvement here. What exactly are you thinking of in terms of strategy? I might be able to look at refactoring this if I have some spare time before the second round of testing.

In the tests (`test_training_progress.py`), some cases were missing as they didn't correspond to any of our existing seeded requirements/involvements. I've now added those cases in for completeness.

Reply:
@elichad This is the strategy design pattern: https://refactoring.guru/design-patterns/strategy

Once applied correctly, this pattern separates behaviours based on some criterium (or criteria). I've been thinking and it may not be the best pattern for us, as there would be multiple input values (url_required, event_required, involvement_required, all with negations too).

Maybe instead we should focus on individual fields and have a validation like so:

```
def validate_url(self, requirement: TrainingRequirement) -> Optional[ValidationError]:
  if self.url and not requirement.url_required and not requirement.involvement_required:
    return ValidationError("asd")
  if not self.url and requirement.url_required:
    return ValidationError("asd1")
  if requirement.involvement_required and self.involvement_type and self.involvement_type.url_required and not self.url:
    return ValidationError("asd2")


def validate_event(self, requirement: TrainingRequirement) -> Optional[ValidationError]:
  if self.event and not requirement.event_required:
    return ValidationError("asd")
  if not self.event and requirement.event_required:
    return ValidationError("dsa")


def validate_involvement_type(self, requirement: TrainingRequirement) -> Optional[ValidationError]:
  if self.involvement_type and not requirement.involvement_required:
    return ValidationError("asd")
  if not self.involvement_type and requirement.involvement_required:
    return ValidationError("dsa")


def validate_date(self, requirement: TrainingRequirement) -> Optional[ValidationError]:
  if not requirement.involvement_required and self.date:
    return ValidationError("asd")
  if requirement.involvement_required and self.involvement_type and self.involvement_type.date_required and not self.date:
    return ValidationError("asd1")
  if requirement.involvement_required and self.involvement_type and not self.involvement_type.date_required and self.date:
    return ValidationError("asd2")
  if requirement.involvement_required and self.involvement_type and self.date and self.date > timezone.localdate(timezone=pytz.timezone("Etc/GMT-14")):
    return ValidationError("asd3")

def validate_notes(self, requirement: TrainingRequirement) -> Optional[ValidationError]:
  ...


def clean(self):
  super().clean()
  errors = {}
  validators = [
    ("url", self.validate_url),
    ("event", self.validate_event),
    ("involvement_type", self.validate_involvement_type),
    ("date", self.validate_date),
    ("notes", self.validate_notes),
  ]

  for field, validator in validators:
    error = validator(self.requirement)
    if error:
      errors[field] = error

  if errors:
    raise ValidationError(errors)
```

Reply:
Thanks. I've moved this to issue #2451